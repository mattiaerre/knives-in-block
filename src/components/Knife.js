import { DOWN, UP } from '../models/constants';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './Knife.module.css';
import { useState } from 'react';

export default function Knife({ color, dispatch, id }) {
  const [state, setState] = useState(DOWN);

  function handleOnClick() {
    if (state === DOWN) {
      setState(UP);
      return dispatch({ type: 'knife_pick_up', value: id });
    }
    if (state === UP) {
      setState(DOWN);
      return dispatch({ type: 'knife_put_down', value: null });
    }
  }

  return (
    <div
      className={classNames(styles[state], styles.knife)}
      data-testid="knife"
      onClick={handleOnClick}
      style={{ backgroundColor: color }}
    ></div>
  );
}

Knife.propTypes = {
  color: PropTypes.string.isRequired,
  dispatch: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired
};
