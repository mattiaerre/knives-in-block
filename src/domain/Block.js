export const empty = 'empty';

export const emptyBlock = [
  [empty, empty, empty],
  [empty, empty, empty]
];

export function Block(initialBlock = emptyBlock) {
  // it's a 2x3 matrix
  const _block = initialBlock;

  return {
    // query
    current: () => {
      return _block;
    },
    // command
    place: (knife, x, y) => {
      _block[x][y] = knife;
    },
    // command
    remove: (x, y) => {
      _block[x][y] = empty;
    }
  };
}
