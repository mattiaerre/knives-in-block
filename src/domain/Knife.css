.down {
}

.up {
  opacity: 50%;
}
