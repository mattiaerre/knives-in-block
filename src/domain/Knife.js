export const blue = 'blue';
export const turquoise = 'turquoise';
export const white = 'white';
