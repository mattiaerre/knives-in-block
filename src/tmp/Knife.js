import { useState } from 'react';
import classNames from 'classnames';

export const blue = 'blue';
export const turquoise = 'turquoise';
export const white = 'white';

export function Knife({ handleOnClickKnife, isSelected, knife }) {
  const [selected, setSelected] = useState(isSelected);

  function handleOnClick(e) {
    setSelected(!selected);
    handleOnClickKnife(e);
  }

  return (
    <li
      className={classNames({ disabled: selected }, 'knife', knife)}
      onClick={handleOnClick}
    />
  );
}

// A knife can be put down, picked up, and inserted into the block
// State machine: Down <-> Up <-> Block
