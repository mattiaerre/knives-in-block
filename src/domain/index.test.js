import { empty, Block } from './Block';
import { blue, turquoise, white } from './Knife';

// given when then
// arrange act assert
// given an empty block place the first knife

test('snapshot', () => {
  const block = Block();
  block.place(turquoise, 1, 2);

  expect(block.current()).toMatchSnapshot();
});

test('place white', () => {
  const block = Block();
  block.place(white, 0, 0);

  expect(block.current()[0][0]).toEqual(white);
});

test('place blue', () => {
  const block = Block();
  block.place(blue, 0, 0);

  expect(block.current()[0][0]).toEqual(blue);
});

test('remove white', () => {
  const block = Block();
  block.place(white, 0, 1);
  block.remove(0, 1);

  expect(block.current()[0][1]).toEqual(empty);
});
