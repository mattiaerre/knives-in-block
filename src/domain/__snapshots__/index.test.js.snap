// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Array [
  Array [
    "empty",
    "empty",
    "empty",
  ],
  Array [
    "empty",
    "empty",
    "turquoise",
  ],
]
`;
